Doppler.BillingUser/Controllers/BillingController.cs

```diff
@@ -119,13 +131,77 @@ public async Task<IActionResult> UpdateCurrentPaymentMethod(string accountname,
         [HttpPost("/accounts/{accountname}/agreements")]
         public async Task<IActionResult> CreateAgreement([FromRoute] string accountname, [FromBody] AgreementInformation agreementInformation)
         {
+            var results = await _agreementInformationValidator.ValidateAsync(agreementInformation);
+            if (!results.IsValid)
+            {
+                return new BadRequestObjectResult(results.ToString("-"));
+            }
+
+            var user = await _userRepository.GetUserBillingInformation(accountname);
+            if (user == null)
+            {
+                return new NotFoundObjectResult("Invalid user");
+            }
+
+            if (user.PaymentMethod != PaymentMethodEnum.CC)
+            {
+                return new BadRequestObjectResult("Invalid payment method");
+            }
+
             var isValidTotal = await _accountPlansService.IsValidTotal(accountname, agreementInformation);
 
             if (!isValidTotal)
             {
                 return new BadRequestObjectResult("Total of agreement is not valid");
             }
 
+            var currentPlan = await _userRepository.GetUserCurrentTypePlan(user.IdUser);
+            if (currentPlan != null)
+            {
+                return new BadRequestObjectResult("Invalid user type (only free users)");
+            }
+
+            var newPlan = await _userRepository.GetUserNewTypePlan(agreementInformation.PlanId);
+            if (newPlan == null)
+            {
+                return new BadRequestObjectResult("Invalid selected plan");
+            }
+
+            if (newPlan.IdUserType != UserTypeEnum.INDIVIDUAL)
+            {
+                return new BadRequestObjectResult("Invalid selected plan type");
+            }
+
+            if (agreementInformation.Total.GetValueOrDefault() > 0)
+            {
+                var encryptedCreditCard = await _userRepository.GetEncryptedCreditCard(accountname);
+                if (encryptedCreditCard == null)
+                {
+                    return new ObjectResult("User credit card missing")
+                    {
+                        StatusCode = 500
+                    };
+                }
+
+                // TODO: Deal with first data exceptions.
+                var authorizationNumber = await _paymentGateway.CreateCreditCardPayment(agreementInformation.Total.GetValueOrDefault(), encryptedCreditCard, user.IdUser);
+                await _billingRepository.CreateAccountingEntriesAsync(agreementInformation, encryptedCreditCard, user.IdUser, authorizationNumber);
+            }
+
+            var billingCreditId = await _billingRepository.CreateBillingCreditAsync(agreementInformation, user, newPlan);
+
+            user.IdCurrentBillingCredit = billingCreditId;
+            await _userRepository.UpdateUserBillingCredit(user);
+
+            if (agreementInformation.Total.HasValue)
```

Comment:
Me parece que lo que querés comprobar no es que tenga valor, sino que el valor sea mayor que cero.

```suggestion
            if (agreementInformation.Total.GetValueOrDefault() > 0)
```

Antes de arreglarlo, estaría bueno hacer tests que cubran este escenario:

* Validar que cuando el valor sea mayor que cero se haga todo (creo que está hecho)
* Validar que cuando el valor sea cero determinadas cosas no se hagan

Reply:
Entiendo, voy a implementar unos test que cubran el escenario.

Reply:
Listo, agregue una prueba, que permite el total cero y valida que no se creen los registros asociados a un pago con tarjeta de crédito.

Reply:
❤️